confluent_kafka/src/get_metadata_impl.c

```diff
@@ -0,0 +1,385 @@
+#include "confluent_kafka.h"
+
+#if RD_KAFKA_VERSION >= 0x00080400
+
+static int
+set_long_in_dict(PyObject *target, const char *key, const long val)
+{
+	int err;
+	PyObject *py_val;
+
+	if ((py_val = PyLong_FromLong(val)) == NULL) {
+		return -1;
+	}
+
+	err =  PyDict_SetItemString(target, key, py_val);
+	Py_DECREF(py_val);
+	return err;
+}
+
+
+static int
+set_char_in_dict(PyObject *target, const char *key, const char *val)
+{
+	int err;
+	PyObject *py_val;
+
+	py_val = PyUnicode_FromString(val);
+
+	if (py_val == NULL) {
+		return -1;
+	}
+
+	err = PyDict_SetItemString(target, key, py_val);
+	Py_DECREF(py_val);
+	return err;
+}
+
+
+#define SET_LONG_IN_DICT(target, ptr, member) \
+    set_long_in_dict(target, #member, (ptr)->member)
+
+
+#define SET_CHAR_IN_DICT(target, ptr, member) \
+    set_char_in_dict(target, #member, (ptr)->member)
+
+struct broker_entry {
```

Comment:
I'm not sure this rich broker_entry is really that valuable, providing the full broker object in each reference to a broker.

With thousands of topics and their partitions we'll have a huge number of objects that are identical.

I'd rather see a top-level dict `brokers` dict indexed by the brokerid where the value is {id, name}.
For all other references to the broker we simply provide the brokerId. It is simple enough for the application to lookup the broker name from the top-level dict and it avoids the en masse object creation which is really slow in Python.

So: remove all broker_entry-related stuff, and have the broker_list thing simply be a list of brokerId ints.

Reply:
are you sure this creates dupplicate objects for each broker? I tried to set things up so that each reference to a broker would have a reference to the same broker object. That's the reason for first making the brokers then doing the find_broker() stuff to get the PyObject increfing it and adding it to another object.  I don't terribly object to your approach, but it is more awkward for consumers and departs from the way java does things, but I can imagine it would be faster and make some operations like listing all brokers easier.